File: examples/opencv-pipeline/reference-opencv/opencv_pipeline.cpp
// SPDX-License-Identifier: MIT

#include <opencv2/imgcodecs.hpp>
#include <opencv2/imgproc.hpp>

#include <chrono>
#include <filesystem>
#include <fstream>
#include <iostream>
#include <stdexcept>
#include <string>

namespace {

struct Config {
    std::filesystem::path input;
    std::filesystem::path out = ".tmp/opencv-pipeline/opencv";
    double sigma = 1.5;
    int min_size = 20;
    std::string mode = "segment";
};

struct StageTimings {
    double transform_seconds = 0.0;
    double smooth_seconds = 0.0;
    double threshold_seconds = 0.0;
    double morphology_seconds = 0.0;
    double filter_seconds = 0.0;
    double label_seconds = 0.0;
    double measure_seconds = 0.0;
};

std::string value(int argc, char** argv, int& i, const char* name) {
    ++i;
    if (i >= argc) {
        throw std::invalid_argument(std::string(name) + " needs a value");
    }
    return argv[i];
}

Config parse(int argc, char** argv) {
    Config config;
    for (int i = 1; i < argc; ++i) {
        std::string arg = argv[i];
        if (arg == "--input") {
            config.input = value(argc, argv, i, "--input");
        } else if (arg == "--out") {
            config.out = value(argc, argv, i, "--out");
        } else if (arg == "--sigma") {
            config.sigma = std::stod(value(argc, argv, i, "--sigma"));
        } else if (arg == "--min-size") {
            config.min_size = std::stoi(value(argc, argv, i, "--min-size"));
        } else if (arg == "--mode") {
            config.mode = value(argc, argv, i, "--mode");
        } else if (arg == "--help" || arg == "-h") {
            std::cout << "opencv_pipeline --input IMAGE --out DIR [--mode segment|transform]\n";
            std::exit(0);
        } else {
            throw std::invalid_argument("unknown argument " + arg);
        }
    }
    if (config.input.empty()) {
        throw std::invalid_argument("missing required --input");
    }
    if (!(config.sigma >= 0.0)) {
        throw std::invalid_argument("--sigma must be non-negative");
    }
    if (config.min_size < 1) {
        throw std::invalid_argument("--min-size must be at least 1");
    }
    if (config.mode != "segment" && config.mode != "transform") {
        throw std::invalid_argument("--mode must be segment or transform");
    }
    return config;
}

double seconds_since(std::chrono::steady_clock::time_point start) {
    auto elapsed = std::chrono::steady_clock::now() - start;
    return std::chrono::duration<double>(elapsed).count();
}

cv::Mat transform_if_requested(const cv::Mat& input, const std::string& mode) {
    if (mode == "segment") {
        return input.clone();
    }
    cv::Mat matrix = (cv::Mat_<double>(2, 3) << 1.0, 0.0, 7.0, 0.0, 1.0, 5.0);
    cv::Mat out;
    cv::warpAffine(input, out, matrix, input.size(), cv::INTER_NEAREST, cv::BORDER_CONSTANT, cv::Scalar(0));
    return out;
}

void write_objects(const cv::Mat& stats, const cv::Mat& centroids, const std::filesystem::path& path) {
    std::ofstream out(path);
    if (!out) {
        throw std::runtime_error("create CSV failed: " + path.string());
    }
    out << "label,count,centroid_y,centroid_x\n";
    for (int label = 1; label < stats.rows; ++label) {
        int area = stats.at<int>(label, cv::CC_STAT_AREA);
        if (area == 0) {
            continue;
        }
        out << label << ',' << area << ',' << centroids.at<double>(label, 1) << ','
            << centroids.at<double>(label, 0) << '\n';
    }
}

void write_summary(const Config& config, int objects, long long total_area, double threshold,
                   double load_seconds, double pipeline_seconds, const StageTimings& timings,
                   const std::filesystem::path& path) {
    std::ofstream out(path);
    if (!out) {
        throw std::runtime_error("create summary failed: " + path.string());
    }
    out << "{\n";
    out << "  \"input\": \"" << config.input.string() << "\",\n";
    out << "  \"mode\": \"" << config.mode << "\",\n";
    out << "  \"objects\": " << objects << ",\n";
    out << "  \"total_foreground_area\": " << total_area << ",\n";
    out << "  \"threshold\": " << threshold << ",\n";
    out << "  \"sigma\": " << config.sigma << ",\n";
    out << "  \"min_size\": " << config.min_size << ",\n";
    out << "  \"load_seconds\": " << load_seconds << ",\n";
    out << "  \"pipeline_seconds\": " << pipeline_seconds << ",\n";
    out << "  \"stage_seconds\": {\n";
    out << "    \"transform\": " << timings.transform_seconds << ",\n";
    out << "    \"smooth\": " << timings.smooth_seconds << ",\n";
    out << "    \"threshold\": " << timings.threshold_seconds << ",\n";
    out << "    \"morphology\": " << timings.morphology_seconds << ",\n";
    out << "    \"filter\": " << timings.filter_seconds << ",\n";
    out << "    \"label\": " << timings.label_seconds << ",\n";
    out << "    \"measure\": " << timings.measure_seconds << "\n";
    out << "  }\n";
    out << "}\n";
}

}  // namespace

int main(int argc, char** argv) {
    try {
        Config config = parse(argc, argv);
        std::filesystem::create_directories(config.out);

        auto started = std::chrono::steady_clock::now();
        cv::Mat input = cv::imread(config.input.string(), cv::IMREAD_GRAYSCALE);
        if (input.empty()) {
            throw std::runtime_error("could not decode " + config.input.string());
        }
        double load_seconds = seconds_since(started);

        auto pipeline_started = std::chrono::steady_clock::now();
        StageTimings timings;
        auto stage_started = std::chrono::steady_clock::now();
        cv::Mat prepared = transform_if_requested(input, config.mode);
        timings.transform_seconds = seconds_since(stage_started);
        stage_started = std::chrono::steady_clock::now();
        cv::Mat smoothed;
        if (config.sigma == 0.0) {
            smoothed = prepared;
        } else {
            cv::GaussianBlur(prepared, smoothed, cv::Size(), config.sigma, config.sigma, cv::BORDER_REFLECT);
        }
        timings.smooth_seconds = seconds_since(stage_started);

        stage_started = std::chrono::steady_clock::now();
        cv::Mat raw_mask;
        double threshold = cv::threshold(smoothed, raw_mask, 0.0, 255.0, cv::THRESH_BINARY | cv::THRESH_OTSU);
        timings.threshold_seconds = seconds_since(stage_started);
        stage_started = std::chrono::steady_clock::now();
        cv::Mat morphed;
        cv::Mat kernel = cv::getStructuringElement(cv::MORPH_RECT, cv::Size(3, 3));
        cv::morphologyEx(raw_mask, morphed, cv::MORPH_OPEN, kernel, cv::Point(-1, -1), 1, cv::BORDER_CONSTANT, cv::Scalar(0));
        cv::morphologyEx(morphed, morphed, cv::MORPH_CLOSE, kernel, cv::Point(-1, -1), 1, cv::BORDER_CONSTANT, cv::Scalar(0));
        timings.morphology_seconds = seconds_since(stage_started);

        stage_started = std::chrono::steady_clock::now();
        cv::Mat labels;
        cv::Mat stats;
        cv::Mat centroids;
        int components = cv::connectedComponentsWithStats(morphed, labels, stats, centroids, 4);
        cv::Mat kept = cv::Mat::zeros(morphed.size(), CV_8U);
        for (int label = 1; label < components; ++label) {
            int area = stats.at<int>(label, cv::CC_STAT_AREA);
            if (area >= config.min_size) {
                kept.setTo(255, labels == label);
            }
        }
        timings.filter_seconds = seconds_since(stage_started);

        stage_started = std::chrono::steady_clock::now();
        components = cv::connectedComponentsWithStats(kept, labels, stats, centroids, 4);
        timings.label_seconds = seconds_since(stage_started);
        stage_started = std::chrono::steady_clock::now();
        int objects = components - 1;
        long long total_area = 0;
        for (int label = 1; label < components; ++label) {
            total_area += stats.at<int>(label, cv::CC_STAT_AREA);
        }
        timings.measure_seconds = seconds_since(stage_started);
        double pipeline_seconds = seconds_since(pipeline_started);

        write_objects(stats, centroids, config.out / "objects.csv");
        write_summary(config, objects, total_area, threshold, load_seconds, pipeline_seconds, timings, config.out / "summary.json");
        std::cout << "objects=" << objects << " threshold=" << threshold << " output=" << config.out << "\n";
        return 0;
    } catch (const std::exception& err) {
        std::cerr << "opencv-pipeline: " << err.what() << "\n";
        return 1;
    }
}

File: examples/opencv-pipeline/src/bin/opencv_pipeline.rs
// SPDX-License-Identifier: MIT

use std::fs::{self, File};
use std::io::{BufWriter, Write};
use std::path::{Path, PathBuf};
use std::time::Instant;

use blockflow::assemble::{ImageId, PlanBuilder};
use blockflow::dtype::Dtype;
use blockflow::geometry::BlockGrid;
use blockflow::op::Chain;
use blockflow::ops::label::LabelComponentsOp;
use blockflow::ops::{
    append_global_threshold_phases, append_remove_small_objects_phases, append_warp_phase,
    collect_class_a_shapes, Boundary, Connectivity, ElementShape, Gaussian, GlobalThreshold,
    GlobalThresholdOutput, GlobalThresholdSelection, Measurements, Morphology, MorphologyOp,
    ShapeMeasurements, ShapeSet, SmoothOp, StructuringElement, ThresholdTest, TransformBoundary,
    TransformInterpolation, WarpOp,
};
use blockflow::sidecar::Lifecycle;
use blockflow::strategy::{execute_phases, Hints};
use blockflow::voxels::Voxels;
use blockflow::zarr_env::ZarrEnvironment;
use blockflow::{AttachedImage, Error, Result};
use clap::{Parser, ValueEnum};
use ndarray::Array3;

#[derive(Debug, Parser)]
#[command(name = "opencv-pipeline")]
struct Config {
    #[arg(long)]
    input: PathBuf,
    #[arg(long, default_value = ".tmp/opencv-pipeline/blockflow")]
    out: PathBuf,
    #[arg(long, default_value_t = 1.5)]
    sigma: f64,
    #[arg(long, default_value_t = 20)]
    min_size: u64,
    #[arg(long, value_enum, default_value = "segment")]
    mode: Mode,
    #[arg(long, default_value = ".tmp/opencv-pipeline/input.zarr")]
    zarr_dir: PathBuf,
    #[arg(long, value_parser = parse_chunk, default_value = "1x256x256")]
    chunk: [usize; 3],
}

#[derive(Debug, Clone, Copy, Eq, PartialEq, ValueEnum)]
enum Mode {
    Segment,
    Transform,
}

impl Mode {
    fn as_str(self) -> &'static str {
        match self {
            Self::Segment => "segment",
            Self::Transform => "transform",
        }
    }
}

#[derive(Debug)]
struct ObjectRow {
    label: u64,
    count: u64,
    centroid_y: f64,
    centroid_x: f64,
}

fn main() {
    if let Err(err) = run() {
        eprintln!("{err}");
        std::process::exit(1);
    }
}

fn run() -> Result<()> {
    let config = Config::parse()?;
    fs::create_dir_all(&config.out)
        .map_err(|err| Error::invalid(format!("opencv-pipeline: create output dir: {err}")))?;

    let started = Instant::now();
    let input_zarr = ensure_input_zarr(&config)?;
    let load_seconds = started.elapsed().as_secs_f64();

    let pipeline_started = Instant::now();
    let rows = planned_pipeline(&config, &input_zarr)?;
    let pipeline_seconds = pipeline_started.elapsed().as_secs_f64();

    write_objects(&rows, &config.out.join("objects.csv"))?;
    write_summary(
        &config,
        rows.len(),
        rows.iter().map(|row| row.count).sum(),
        load_seconds,
        pipeline_seconds,
        &config.out.join("summary.json"),
    )?;

    println!(
        "objects={} threshold_method=otsu output={}",
        rows.len(),
        config.out.display()
    );
    Ok(())
}

impl Config {
    fn parse() -> Result<Self> {
        let config = <Self as Parser>::parse();

        if config.sigma < 0.0 || !config.sigma.is_finite() {
            return Err(Error::invalid(
                "opencv-pipeline: --sigma must be finite and non-negative",
            ));
        }
        if config.min_size == 0 {
            return Err(Error::invalid(
                "opencv-pipeline: --min-size must be at least 1",
            ));
        }
        if config.chunk.contains(&0) {
            return Err(Error::invalid(
                "opencv-pipeline: --chunk dimensions must be positive",
            ));
        }

        Ok(config)
    }
}

fn ensure_input_zarr(config: &Config) -> Result<PathBuf> {
    let array = config.zarr_dir.join("level0");
    if array.join("zarr.json").exists() {
        let (dtype, _) = AttachedImage::at(&array).metadata()?;
        if dtype != Dtype::F64 {
            return Err(Error::invalid(format!(
                "opencv-pipeline: prepared input {} is {dtype:?}, expected F64",
                array.display()
            )));
        }
        return Ok(array);
    }

    let input = load_luma(&config.input)?;
    let voxels: Voxels = input.into();
    ZarrEnvironment::create(&config.zarr_dir, &voxels, config.chunk)?;
    Ok(array)
}

fn planned_pipeline(config: &Config, input_zarr: &Path) -> Result<Vec<ObjectRow>> {
    let (_, volume) = AttachedImage::at(input_zarr).metadata()?;
    let grid = BlockGrid::new(volume, config.chunk)?;
    let mut builder = PlanBuilder::new(volume, Dtype::F64, grid.clone());
    if config.mode == Mode::Transform {
        append_warp_phase(
            &mut builder,
            WarpOp::affine(
                "opencv-translate",
                [[1.0, 0.0, 0.0], [0.0, 1.0, 0.0], [0.0, 0.0, 1.0]],
                [0.0, -5.0, -7.0],
                volume,
                TransformInterpolation::Linear,
                TransformBoundary::Constant(0.0),
            )?,
            volume,
            grid,
        )?;
    }
    builder.pixels(Chain::op(SmoothOp::new(
        "opencv-gaussian-smooth",
        Gaussian::new([0.0, config.sigma, config.sigma], 3.0)?.with_boundary(Boundary::Reflect),
    )))?;
    append_global_threshold_phases(
        &mut builder,
        "opencv-threshold",
        GlobalThresholdSelection::single(GlobalThreshold::otsu(256)?),
        GlobalThresholdOutput::Mask {
            test: ThresholdTest::Above,
        },
    )?;
    let element = StructuringElement::from_radius(ElementShape::Box, [0, 1, 1]);
    builder.pixels(Chain::op(MorphologyOp::new(
        "opencv-open-3x3",
        Morphology::Open,
        element.clone(),
    )))?;
    builder.pixels(Chain::op(MorphologyOp::new(
        "opencv-close-3x3",
        Morphology::Close,
        element,
    )))?;
    append_remove_small_objects_phases(
        &mut builder,
        "opencv-remove-small",
        Lifecycle::DeleteOnExit,
        Connectivity::Faces,
        config.min_size,
    )?;
    builder.fragments(
        LabelComponentsOp::new(
            "opencv-label-components",
            "opencv.components",
            Lifecycle::DeleteOnExit,
        )
        .connecting(Connectivity::Faces),
    )?;
    let base = builder.finish()?;
    let labels = ImageId::from(base.n_phases());
    let measurements = Measurements::for_labels(labels)
        .shape(ShapeSet::basic())
        .stream("opencv.measurements")
        .lifecycle(Lifecycle::DeleteOnExit)
        .build(base.decomposition.clone())?;
    let shape_rows = measurements
        .class_a_rows()
        .ok_or_else(|| Error::invalid("opencv-pipeline: planned shape rows are missing"))?;
    let env = ZarrEnvironment::attach(
        config.zarr_dir.join("work.zarr"),
        &[AttachedImage::at(input_zarr)],
    )?;
    let mut work = base.work();
    work.extend(measurements.phase_work());
    execute_phases(
        "opencv planned pipeline",
        &base.workflow,
        &measurements.decomposition,
        &Hints::default(),
        &env,
        &[],
        &work,
    )?;
    let mut rows = collect_class_a_shapes(&env, &shape_rows, volume, measurements.fixed)?
        .into_iter()
        .map(|shape| {
            let shape = ShapeMeasurements::from_shape(&shape);
            let centroid = shape.centroid.unwrap_or([f64::NAN; 3]);
            ObjectRow {
                label: shape.label,
                count: shape.count,
                centroid_y: centroid[1],
                centroid_x: centroid[2],
            }
        })
        .collect::<Vec<_>>();
    rows.sort_by_key(|row| row.label);
    Ok(rows)
}

fn load_luma(path: &Path) -> Result<Array3<f64>> {
    let image = image::ImageReader::open(path)
        .map_err(|err| Error::invalid(format!("opencv-pipeline: open image: {err}")))?
        .decode()
        .map_err(|err| Error::invalid(format!("opencv-pipeline: decode image: {err}")))?
        .to_luma8();
    let (width, height) = image.dimensions();
    let width = usize::try_from(width)
        .map_err(|_| Error::invalid("opencv-pipeline: image width does not fit usize"))?;
    let height = usize::try_from(height)
        .map_err(|_| Error::invalid("opencv-pipeline: image height does not fit usize"))?;
    let mut out = Array3::<f64>::zeros((1, height, width));
    for (x, y, pixel) in image.enumerate_pixels() {
        out[[0, y as usize, x as usize]] = f64::from(pixel.0[0]);
    }
    Ok(out)
}

fn write_objects(rows: &[ObjectRow], path: &Path) -> Result<()> {
    let file = File::create(path)
        .map_err(|err| Error::invalid(format!("opencv-pipeline: create CSV: {err}")))?;
    let mut out = BufWriter::new(file);
    writeln!(out, "label,count,centroid_y,centroid_x").map_err(write_error)?;
    for row in rows {
        writeln!(
            out,
            "{},{},{},{}",
            row.label, row.count, row.centroid_y, row.centroid_x
        )
        .map_err(write_error)?;
    }
    out.flush().map_err(write_error)
}

fn write_summary(
    config: &Config,
    objects: usize,
    total_area: u64,
    load_seconds: f64,
    pipeline_seconds: f64,
    path: &Path,
) -> Result<()> {
    let summary = serde_json::json!({
        "input": config.input,
        "mode": config.mode.as_str(),
        "objects": objects,
        "total_foreground_area": total_area,
        "threshold_method": "otsu",
        "sigma": config.sigma,
        "min_size": config.min_size,
        "load_seconds": load_seconds,
        "pipeline_seconds": pipeline_seconds,
        "input_zarr": config.zarr_dir.join("level0"),
        "chunk_shape": config.chunk,
        "execution": "planned segmentation over attached Zarr input",
    });
    let text = serde_json::to_string_pretty(&summary)
        .map_err(|err| Error::invalid(format!("opencv-pipeline: encode summary: {err}")))?;
    fs::write(path, text)
        .map_err(|err| Error::invalid(format!("opencv-pipeline: write summary: {err}")))
}

fn write_error(err: std::io::Error) -> Error {
    Error::invalid(format!("opencv-pipeline: write output: {err}"))
}

fn parse_chunk(raw: &str) -> std::result::Result<[usize; 3], String> {
    let parts = raw
        .split(['x', 'X', ',', ':'])
        .map(str::trim)
        .collect::<Vec<_>>();
    if parts.len() != 3 {
        return Err(format!(
            "opencv-pipeline: chunk shape {raw:?} must have three dimensions"
        ));
    }
    let mut out = [0usize; 3];
    for (index, part) in parts.iter().enumerate() {
        out[index] = part.parse::<usize>().map_err(|err| {
            format!("opencv-pipeline: could not parse chunk shape {raw:?}: {err}")
        })?;
    }
    Ok(out)
}
